src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/ProjectReloadManager.cs

```diff
@@ -0,0 +1,508 @@
+            {
+                await _threadHandling.SwitchToUIThread();
+
+                IVsSolution solution = _serviceProvider.GetService<IVsSolution, SVsSolution>();
+                if (solution != null)
+                {
+                    int hr = solution.UnadviseSolutionEvents(_solutionEventsCookie);
+                    System.Diagnostics.Debug.Assert(ErrorHandler.Succeeded(hr));
+                    _solutionEventsCookie = VSConstants.VSCOOKIE_NIL;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Handles one time initialization
+        /// </summary>
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            _reloadDelayScheduler = new TaskDelayScheduler(TimeSpan.FromMilliseconds(ReloadDelay), _threadHandling, CancellationToken.None);
+            await ConnectToSolutionEvents().ConfigureAwait(false);
+
+        }
+
+        /// <summary>
+        /// IDisposable handler. Should only be called once
+        /// </summary>
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            if(_reloadDelayScheduler != null)
+            {
+                _reloadDelayScheduler.Dispose();
+                _reloadDelayScheduler = null;
+            }
+            await DisconnectFromSolutionEvents().ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Called when one of the project files changes. In our case since only one file is watched with each cookie so the list of files
+        /// should be one.
+        /// </summary>
+        public int FilesChanged(uint cChanges, string[] rgpszFile, uint[] grfChange)
+        {
+            if(cChanges == 1 && (grfChange[0] & (uint)(_VSFILECHANGEFLAGS.VSFILECHG_Size | _VSFILECHANGEFLAGS.VSFILECHG_Time)) != 0)
+            {
+                lock(_changedProjects)
+                {
+                    var changedProject = _registeredProjects.FirstOrDefault(kv => kv.Key.ProjectFile.Equals(rgpszFile[0], StringComparison.OrdinalIgnoreCase)).Key;
+                    if(changedProject != null)
+                    {
+                        if(!_changedProjects.Contains(changedProject))
+                        {
+                            _changedProjects.Add(changedProject);
+                        }
+                        _reloadDelayScheduler.ScheduleAsyncTask(async (ct) => 
+                        {
+                            // Grab the UI thread so that we block until the reload of this set of 
+                            // projects completes.
+                            await _threadHandling.SwitchToUIThread();
+
+                            // Make a copy of the changed projects
+                            var changedProjects = new List<IReloadableProject>();
+                            lock(_changedProjects)
+                            {
+                                changedProjects.AddRange(_changedProjects);
```

Comment:
Thinking along the lines of the hundreds of projects scenario (perhaps a git checkout changed a bunch at once), instead of doing a mass copy, might we instead just create a new list and swap it for the existing one? Updating the reference should be much quicker than copying a whole bunch of projects.